### PR TITLE
Remove BinarySearch from Adam Op (CPU part)

### PR DESCRIPTION
**Pull request by velconia**
针对CPU情况，假设参数大小为 N, 梯度大小为M, 则原有算法复杂度为 O(N * log(M))
修改后，复杂度为: O(M * log(M) + N);

(默认 N >>> M)

this is for #14861

### CODE REVIEW

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 1]
```diff
@@ -236,6 +239,117 @@ struct SparseAdamFunctor {
   }
 };
 
+template <typename T>
+struct SparseAdamFunctor<T, CPUAdam> {
```
Comment (jacquesqiao) [inline paddle/fluid/operators/optimizers/adam_op.h:243]: Can this functor be merged to SparseAdamFunctor?

Comment (velconia) [reply paddle/fluid/operators/optimizers/adam_op.h:243]: Good advice, GPU part will be fixed in the follow ups

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 2]
```diff
@@ -236,6 +239,117 @@ struct SparseAdamFunctor {
...
+    T lr = *lr_;
+    T beta1_pow = *beta1_pow_;
+    T beta2_pow = *beta2_pow_;
+    lr *= sqrt(1 - beta2_pow) / (1 - beta1_pow);
+    size_t row_count = numel / row_numel_;
+
+    for (size_t i = 0U, j = 0U; i != row_count; ++i) {
+      if (i == *(rows_ + j)) {
+        for (size_t k = 0U; k != row_numel_; ++k) {
+          T mom1 = moment1_[i * row_numel_ + k];
```
Comment (jacquesqiao) [inline paddle/fluid/operators/optimizers/adam_op.h]: can call 
```cpp
inline HOSTDEVICE void adam_update(size_t i, T g) const {}
```
directly

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 3]
```diff
@@ -236,6 +239,117 @@ struct SparseAdamFunctor {
...
+          p -= lr * (mom1 / (sqrt(mom2) + epsilon_));
+          // Write back to global memory
+          moment1_out_[i * row_numel_ + k] = mom1;
+          moment2_out_[i * row_numel_ + k] = mom2;
+          param_out_[i * row_numel_ + k] = p;
+        }
+        ++j;
+      } else {
+        for (size_t k = 0U; k != row_numel_; ++k) {
+          T mom1 = moment1_[i * row_numel_ + k];
```
Comment (jacquesqiao) [inline paddle/fluid/operators/optimizers/adam_op.h:324]: can call 
```cpp
inline HOSTDEVICE void adam_update(size_t i, T g) const {}
```
directly

File: paddle/fluid/operators/optimizers/adam_op.h [hunk 4]
```diff
@@ -347,32 +461,46 @@ class AdamOpKernel : public framework::OpKernel<T> {
...
+            beta2_pow.template data<T>(), mom1.template data<T>(),
+            mom1_out.template mutable_data<T>(ctx.GetPlace()),
+            mom2.template data<T>(),
+            mom2_out.template mutable_data<T>(ctx.GetPlace()),
+            lr.template data<T>(), grad_data, param.template data<T>(),
+            param_out.template mutable_data<T>(ctx.GetPlace()), rows, row_numel,
+            grad_merge.rows().size(), lazy_mode);
+
+        // FIXME(minqiyang): remove BinarySearch in GPU later
         platform::ForRange<DeviceContext> for_range(
```
Comment (jacquesqiao) [inline paddle/fluid/operators/optimizers/adam_op.h:493]: GPU should support lazy mode

**Comment by jacquesqiao** [comment]
Please pull the develop code and run test again.

**Review by jacquesqiao — CHANGES_REQUESTED** [review]
Please pull the develop code and run test again.

**Review by jacquesqiao — APPROVED** [review]
LGTM

**Pull request merged**